File: src/main/java/rustleund/fightingfantasy/framework/closures/impl/TestEnemyTypesPredicate.java
package rustleund.fightingfantasy.framework.closures.impl;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;

import org.w3c.dom.Element;

import rustleund.fightingfantasy.framework.base.Enemies;
import rustleund.fightingfantasy.framework.base.EnemyState;
import rustleund.fightingfantasy.framework.base.GameState;

import com.google.common.base.Predicate;

public class TestEnemyTypesPredicate implements Predicate<GameState> {

	private Collection<String> equalAny = new ArrayList<>();
	private boolean alwaysCheckFirst;
	private int enemyId;

	public TestEnemyTypesPredicate(Element element) {
		if (element.hasAttribute("equalAny")) {
			this.equalAny.addAll(Arrays.asList(element.getAttribute("equalAny").split(",")));
		}
		String configuredId = element.getAttribute("enemyId");
		if ("first".equals(configuredId)) {
			this.alwaysCheckFirst = true;
		} else {
			this.enemyId = Integer.valueOf(configuredId);
		}
	}

	@Override
	public boolean apply(GameState input) {
		Enemies enemies = input.getBattleState().getEnemies();
		EnemyState enemyStateToCheck = alwaysCheckFirst ? enemies.getFirstNonDeadEnemy() : enemies.getEnemies().get(enemyId);
		return equalAny.stream().anyMatch(enemyStateToCheck::isOfType);
	}
}

File: src/main/java/rustleund/fightingfantasy/framework/closures/impl/TestEnemyTypesPredicate.kt
package rustleund.fightingfantasy.framework.closures.impl

import org.w3c.dom.Element
import rustleund.fightingfantasy.framework.base.GameState
import rustleund.fightingfantasy.framework.base.EnemyState
import rustleund.fightingfantasy.framework.base.optionalAttribute
import java.util.function.Predicate

class TestEnemyTypesPredicate(element: Element) : Predicate<GameState> {

    private val equalAny = element.optionalAttribute("equalAny")?.split(',').orEmpty()
    private val alwaysCheckAny = element.getAttribute("enemyIdx") == "any"
    private val enemyIdx = element.getAttribute("enemyIdx").takeIf { it != "any" }?.toInt() ?: 0

    override fun test(input: GameState): Boolean {
        val battleState = input.battleState
        return if (battleState != null) {
            val enemies = battleState.enemies
            val enemyIsOfAnyType: (EnemyState) -> Boolean = { e -> equalAny.any { e.isOfType(it) } }
            if (alwaysCheckAny) enemies.enemies.any(enemyIsOfAnyType)
            else enemyIsOfAnyType(enemies.enemies[enemyIdx])
        } else false
    }
}